rfcs/0000-safe-transmute.md

```diff
@@ -0,0 +1,2160 @@
+
+```rust
+pub mod crate_a {
+
+    #[repr(C)]
+    pub struct NonEmptySlice<T> {
+        data: *const T,
+        len: usize,
+    }
+
+    impl<T> NonEmptySlice<T> {
+        pub fn from_array<const N: usize>(arr: &[T; N], len: usize) -> Self {
+            assert!(len <= N);
+            assert!(len > 0);
+            Self {
+                data: arr as *const T,
+                len,
+            }
+        }
+
+        pub fn first(&self) -> &T {
+            unsafe { &*self.data }
+        }
+    }
+
+}
+```
+It is sound for `first` to be a *safe* method is because the `from_array` constructor ensures that `data` is safe to dereference, and because `from_array` is the *only* way to safely initialize `NonEmptySlice` outside of `crate_a` (note that `NonEmptySlice`'s fields are *not* `pub`). As a rule: any field that is not marked `pub` should be assumed to be private *because* it is subject to safety invariants.
+
+Unfortunately, field visibility modifiers are not a surefire indicator of whether a type is *fully* implicitly constructable. A type author may restrict the implicit constructability of a type even in situations where all fields of that type (*and all fields of those fields*) are `pub`; consider:
+```rust
+pub mod crate_a {
+
+    #[repr(C)]
+    pub struct NonEmptySlice<T>(pub private::NonEmptySliceInner<T>);
+
+    impl<T> NonEmptySlice<T> {
+        pub fn from_array<const N: usize>(arr: &[T; N], len: usize) -> Self {
+            assert!(len <= N && len > 0);
+            Self(
+                private::NonEmptySliceInner {
+                    data: arr as *const T,
+                    len,
+                }
+            )
+        }
+
+        pub fn first(&self) -> &T {
+            unsafe { &*self.0.data }
+        }
+    }
+
+    // introduce a private module to avoid `private_in_public` error (E0446):
+    pub(crate) mod private {
+        #[repr(C)]
+        pub struct NonEmptySliceInner<T> {
+            pub data: *const T,
+            pub len: usize,
+        }
+    }
+
+}
+```
+In the above example, the definitions of both `NonEmptySlice` and its field `NonEmptySliceInner` are marked `pub`, and all fields of these types are marked `pub`. However, `NonEmptySlice` is *not* fully implicitly constructible outside of `crate_a`, because the module containing `NonEmptySliceInner` is not visibile outside of `crate_a`.
+
+#### Constructability and Transmutation
+Transmutation supplies a mechanism for constructing instances of a type *without* invoking its implicit constructor, nor any constructors defined by the type's author.
+
+In the previous examples, it would be *unsafe* to transmute `0u128` into `NonEmptySlice` outside `crate_a`, because subsequent *safe* use of that value (namely, calling `first`) would memory safety. (However, it's completely safe to transmute `NonEmptySlice` into a `u128`.)
```

Comment:
```suggestion
In the previous examples, it would be *unsafe* to transmute `0u128` into `NonEmptySlice` outside `crate_a`, because subsequent *safe* use of that value (namely, calling `first`) would violate memory safety. (However, it's completely safe to transmute `NonEmptySlice` into a `u128`.)
```